File: src/main/java/com/ahold/technl/sandbox/validation/ValidDeliveryStatus.java
package com.ahold.technl.sandbox.validation;

import java.lang.annotation.RetentionPolicy;
import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;
@Constraint(validatedBy = DeliveryStatusValidator.class)
@Target({ ElementType.FIELD, ElementType.PARAMETER })
@Retention(RetentionPolicy.RUNTIME)
public @interface ValidDeliveryStatus {
    Class<? extends Enum<?>> enumClass();
    String message() default "Invalid value. Allowed values are: {enumClass}";
    Class<?>[] groups() default {};
    Class<? extends Payload>[] payload() default {};
}

File: src/main/kotlin/com/ahold/technl/sandbox/validator/ValidDeliveryStatus.kt
package com.ahold.technl.sandbox.validator

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@MustBeDocumented
@Constraint(validatedBy = [DeliveryStatusValidator::class])
@Target(AnnotationTarget.FIELD, AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
annotation class ValidDeliveryStatus(
    val message: String = "Invalid status. Allowed values are IN_PROGRESS, DELIVERED",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)
